File: frontend/src/api/plotMode.ts
import type {
  BootstrapState,
  PlotModeExecutionMode,
  PlotModePathSelectionType,
  PlotModePathSuggestionResponse,
  PlotModeState,
} from "../types";
import { fetchJSON } from "./client";

export function activatePlotWorkspace(id: string) {
  return fetchJSON<BootstrapState>("/api/plot-mode/activate", {
    method: "POST",
    body: JSON.stringify({ id }),
  });
}

export function fetchPlotModePathSuggestions(
  workspaceId: string | null,
  query: string,
  selectionType: PlotModePathSelectionType,
) {
  return fetchJSON<PlotModePathSuggestionResponse>("/api/plot-mode/path-suggestions", {
    method: "POST",
    body: JSON.stringify({
      query,
      selection_type: selectionType,
      workspace_id: workspaceId,
    }),
  });
}

export function selectPlotModePaths(
  workspaceId: string | null,
  selectionType: PlotModePathSelectionType,
  paths: string[],
) {
  return fetchJSON<BootstrapState>("/api/plot-mode/select-paths", {
    method: "POST",
    body: JSON.stringify({
      selection_type: selectionType,
      paths,
      workspace_id: workspaceId,
    }),
  });
}

export function sendPlotModeMessage(body: Record<string, string>) {
  return fetchJSON<{
    status: "ok" | "error";
    plot_mode: PlotModeState;
    error?: string;
  }>("/api/plot-mode/chat", {
    method: "POST",
    body: JSON.stringify(body),
  });
}

export function submitPlotModeTabularHint(body: Record<string, unknown>) {
  return fetchJSON<{
    status: "ok";
    plot_mode: PlotModeState;
  }>("/api/plot-mode/tabular-hint", {
    method: "POST",
    body: JSON.stringify(body),
  });
}

export function updatePlotModeSettings(
  workspaceId: string | null,
  executionMode: PlotModeExecutionMode,
) {
  return fetchJSON<{
    status: "ok";
    plot_mode: PlotModeState;
  }>("/api/plot-mode/settings", {
    method: "PATCH",
    body: JSON.stringify({ execution_mode: executionMode, workspace_id: workspaceId }),
  });
}

export function answerPlotModeQuestion(body: Record<string, unknown>) {
  return fetchJSON<{
    status: "ok";
    plot_mode: PlotModeState;
  }>("/api/plot-mode/answer", {
    method: "POST",
    body: JSON.stringify(body),
  });
}

export function finalizePlotMode(
  workspaceId: string | null,
  metadata?: Record<string, string | null>,
) {
  return fetchJSON<BootstrapState>("/api/plot-mode/finalize", {
    method: "POST",
    body: JSON.stringify({ ...(metadata ?? {}), workspace_id: workspaceId }),
  });
}

export function updatePlotWorkspace(id: string, workspaceName: string) {
  return fetchJSON<{
    status: string;
    plot_mode: PlotModeState;
  }>("/api/plot-mode/workspace", {
    method: "PATCH",
    body: JSON.stringify({ id, workspace_name: workspaceName }),
  });
}

export function deletePlotWorkspace(id: string) {
  return fetchJSON<BootstrapState>("/api/plot-mode", {
    method: "DELETE",
    body: JSON.stringify({ id }),
  });
}

File: src/openplot/api/plot_mode.py
from __future__ import annotations

from fastapi import APIRouter, Request

from ..services import plot_mode as plot_mode_service
from .schemas import (
    PlotModeChatRequest,
    PlotModeFinalizeRequest,
    PlotModePathSuggestionsRequest,
    PlotModeQuestionAnswerRequest,
    PlotModeSelectPathsRequest,
    PlotModeSettingsRequest,
    PlotModeTabularHintRequest,
)

router = APIRouter()


@router.get("/api/plot-mode")
async def get_plot_mode_state(request: Request):
    return plot_mode_service.get_plot_mode_state(request.app.state.runtime)


@router.post("/api/plot-mode/files")
async def set_plot_mode_files():
    return await plot_mode_service.set_plot_mode_files()


@router.post("/api/plot-mode/path-suggestions")
async def suggest_plot_mode_paths(
    body: PlotModePathSuggestionsRequest,
    request: Request,
):
    return await plot_mode_service.suggest_plot_mode_paths(
        body, request.app.state.runtime
    )


@router.post("/api/plot-mode/select-paths")
async def select_plot_mode_paths(
    body: PlotModeSelectPathsRequest,
    request: Request,
):
    return await plot_mode_service.select_plot_mode_paths(
        body, request.app.state.runtime
    )


@router.patch("/api/plot-mode/settings")
async def update_plot_mode_settings(body: PlotModeSettingsRequest, request: Request):
    return await plot_mode_service.update_plot_mode_settings(
        body, request.app.state.runtime
    )


@router.post("/api/plot-mode/tabular-hint")
async def submit_plot_mode_tabular_hint(
    body: PlotModeTabularHintRequest, request: Request
):
    return await plot_mode_service.submit_plot_mode_tabular_hint(
        body, request.app.state.runtime
    )


@router.post("/api/plot-mode/answer")
async def answer_plot_mode_question(
    body: PlotModeQuestionAnswerRequest, request: Request
):
    return await plot_mode_service.answer_plot_mode_question(
        body, request.app.state.runtime
    )


@router.post("/api/plot-mode/chat")
async def run_plot_mode_chat(body: PlotModeChatRequest, request: Request):
    return await plot_mode_service.run_plot_mode_chat(body, request.app.state.runtime)


@router.post("/api/plot-mode/finalize")
async def finalize_plot_mode(body: PlotModeFinalizeRequest, request: Request):
    return await plot_mode_service.finalize_plot_mode(body, request.app.state.runtime)


@router.patch("/api/plot-mode/workspace")
async def rename_plot_mode_workspace(request: Request):
    body = await request.json()
    return await plot_mode_service.rename_plot_mode_workspace(
        request.app.state.runtime, body
    )


@router.delete("/api/plot-mode")
async def delete_plot_mode_workspace(request: Request):
    requested_id = None
    try:
        body = await request.json()
        if isinstance(body, dict):
            requested_id = body.get("id")
    except Exception:
        pass
    return await plot_mode_service.delete_plot_mode_workspace(
        request.app.state.runtime,
        requested_id,
    )


@router.post("/api/plot-mode/activate")
async def activate_plot_mode(request: Request):
    requested_id = None
    try:
        body = await request.json()
        if isinstance(body, dict):
            requested_id = body.get("id")
    except Exception:
        pass
    return await plot_mode_service.activate_plot_mode(
        request.app.state.runtime,
        requested_id,
    )
